source-academy/cadet-frontend
# P2Plus: SourceAcademy Mission Creation Interface

### Description
Creates a way for GitHub-hosted missions to be created through the Source Academy website.

Subfeatures:
- Added MCQ support for GitHub-hosted missions
- Ability to create and delete new tasks for a mission
- Ability to create, edit and delete test cases for a mission
- Ability to load test cases
- Control of admin features so that they are only accessible to teachers

### Type of change

### How to test
To test the autotester and MCQ functionality (as a student):
1. Accept the assignment at this link: https://classroom.github.com/a/D16hWvjA
2. Open the mission in the SA interface by deploying this branch.
3. Attempt questions 1 and 2, making use of the autotester to test the results.
4. Attempt questions 3, 4 and 5, saving the chosen answer. Results should be saved to the repository.

To test the edit functionalities:
1. Clone the following repository: https://github.com/githubclassroomtest24032021/sa-tester-test
2. Open the mission in the SA interface by deploying this branch.
3. Additional teacher-only functionality should be available in this branch:
    - the ability to delete the current question
    - the ability to add a new question
    - the ability to delete specific test cases
    - the ability to add new tests cases
    - the ability to delete the 
    - the ability to view and edit the test prepend and test postpend per question
    - the ability to toggle between showing raw text and MCQ display for MCQ questions
3. Use the "add" and "delete" buttons in the control bar to add and delete questions. Save to the repository and check that new questions are created.
4. Click on the task description and the briefing in the side content. Doing so should display the underlying markdown. Make any changes and click away; this should cause the markdown to re-render. Saving to the repository should cause "Problem.md" and "README.md" to change.
5. Attempt to make changes to the code. Saving to the repository should cause edits to "StarterCode.js" and not "SavedCode.js".
6. Navigate to the autotester in the side-content tabs. Add and delete test cases as necessary. Also edit the test programs and expected answers. This should edit or create the file "Testcases.json" in the repository when saved.
7. In the side-content tab, edit the test-prepend and test-postpend. Save to the repository, and see that "TestPrepend.js" and/or "TestPostpend.js" are edited or created.

To test the mission creation functionality:
1. In the Mission Listing page, click on "Create New Mission"
2. Make any changes by
    - Adding questions
    - Editing the briefing
    - Editing the code or description for any question
    - Editing the testcases, test prepend and test postpend for any questions
3. Save the created mission to a new repository.
4. Go back to the mission listing page and load the repository. Check that any saved changes can be loaded.

Issue not fixed:
- Octokit REST API does not allow for repositories to be created for an organisation, only for a user. The ability should exist but this will take a bit of time as we find the new module to add to the project.

### Checklist
